### PR TITLE
Add run_constrained pennylane-lightning <0.0a0.

### PR DESCRIPTION
**Pull request by vincentmr**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -76,11 +82,19 @@ outputs:
...
         - python
-        - setuptools <=64.0.2 # version requirement from cvxpy
+        - setuptools
       run:
         - ninja
         - numpy
         - python
+      run_constrained:
+        - pennylane-lightning <0.0a0
+      test:
```
Comment (leofang) [inline recipe/meta.yaml:92]: Fix indentation level:
```suggestion
    test:
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -53,6 +53,12 @@ outputs:
         - scipy
         - semantic_version >=2.10
         - toml
```
Comment (leofang) [inline recipe/meta.yaml:55]: btw it seems we can also add this to be safe?
```suggestion
        - semantic_version >=2.10
        - toml
      run_constrained:
        - pennylane >=0.29.0
```
This is saying: If `pennylane` is not installed to an environment where `pennylane-core` is installed, it's fine. But if it's installed, it must not clobber with `pennylane-core`, and hence it must be of a recent version after which the splitting (new format) was introduced.

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml (hunk 2 above)
Comment (vincentmr) [reply recipe/meta.yaml:55]: Make sense, thanks.

**Comment by vincentmr** [comment]
I'm getting errors such as 
```
error: no member named 'aligned_alloc' in the global namespace
```
in OSX. @leofang you have not seen this before by any chance?

**Comment by leofang** [comment]
Yes, I saw that in the previous CI run too. That's why I was wondering if rerendering (which could update the CI infra) would fix it, but it's not. I  don't think the package constraint would trigger a compiler error. I seem to recall that someone recently mentioned an osx compiler issue somewhere...  I am poking around.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -76,11 +84,19 @@ outputs:
         - pip
         - pybind11
         - python
-        - setuptools <=64.0.2 # version requirement from cvxpy
+        - setuptools
       run:
         - ninja
         - numpy
         - python
```
Comment (leofang) [inline recipe/meta.yaml:91]: It's nice that the import test exposes this 🙂 It seems `pennylane-core` should be a run requirement for `pennylane-lightning-core`?
```suggestion
        - python
        - {{ pin_subpackage('pennylane-core') }}
```

**Comment by leofang** [comment]
It's better to keep the linter happy, otherwise it'd show up every time we update the recipe.

**Comment by leofang** [comment]
Let's try using a newer OS X SDK to build. Could you add a new file `recipes/conda_build_config.yaml` and add the the following content? (As suggested in the Gitter chat: https://matrix.to/#/!SOyumkgPRWoXfQYIFH:matrix.org/$1635405605328azMuC:gitter.im?via=matrix.org&via=gitter.im&via=cadair.com)
```
MACOSX_SDK_VERSION:        # [osx]
  - "10.14"                # [osx and x86_64]
```
(OS X Arm64 support is a different story, we'll get to that separately...)

File: recipe/meta.yaml (hunk 3 above)
Comment (leofang) [reply recipe/meta.yaml:91]: Yes, and it'd use the `pennylane-core` version seen from the same recipe as the minimal version, so in this case it's equivalent to
```
- pennylane-core >=0.29.0,<1
```
If you don't feel comfortable about `max_pin`, you can adjust it, the doc is here:
https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#extra-jinja2-functions

**Comment by vincentmr** [comment]
@conda-forge-admin, please rerender

**Comment by leofang** [comment]
Linux is back. I'll try to check the OS X build issue locally, it's a bit confusing to me that the same build env was used in #16 (same SDK, same clang) without problem.

**Comment by vincentmr** [comment]
I'm a bit lost here, thanks @leofang . Maybe @BastianZim would have an idea what's going on?

File: recipe/conda_build_config.yaml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+MACOSX_SDK_VERSION:        # [osx]
+  - "10.14"                # [osx and x86_64]
```
Comment (leofang) [inline recipe/conda_build_config.yaml:2]: @vincentmr we were very close! 😅
```suggestion
  - "10.15"                # [osx and x86_64]
```

**Comment by vincentmr** [comment]
@conda-forge-admin, please rerender

**Comment by leofang** [comment]
I think between when #16 was merged and now, only the CI runner image was changed. But from [the change log](https://github.com/actions/runner-images/releases/tag/macOS-11%2F20230318.1) I couldn't tell if the system headers were touched. Anyway, I confirmed in #18 that using macOS SDK 10.15 would work. (Building osx packages could be quite annoying sometimes... 😅)

As promised, let's get M1/M2 supported in a separate PR.

**Review by leofang — APPROVED** [review]

**Comment by vincentmr** [comment]
Thanks again for your help @leofang .

**Comment by vincentmr** [comment]
Hi @leofang , I'm still experiencing the PL version issue you reported [here](https://github.com/conda-forge/pennylane-lightning-gpu-feedstock/issues/1#issuecomment-1494761395). Is it the point of [your PR](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/425)?

**Comment by leofang** [comment]
Yes, that's the hope... Normally conda would look for packages with a higher build number, so if we rebuild we shouldn't need to fix the repodata. But in this case I feel that the multiple "buggy" (under-constrained) packages give conda a lot of headache to resolve it properly (mamba does the right job). So fixing the repodata at least on one side should hopefully push us out of this trap.

**Comment by leofang** [comment]
I am wondering if we should also send a repodata patch for `pennylane`... Conda hju
```
$ conda create -n my_pny_env python=3.10 pennylane
...
  pennylane          conda-forge/noarch::pennylane-0.23.0-pyhd8ed1ab_0 
  pennylane-lightni~ conda-forge/linux-64::pennylane-lightning-0.24.0-py310h3839ac3_0 
...
```
We can certainly add a version constraint like I mentioned [elsewhere](https://github.com/conda-forge/pennylane-lightning-gpu-feedstock/issues/1#issuecomment-1496089447):
```
$ conda create -n my_pny_env python=3.10 "pennylane>=0.29"
...
  pennylane          conda-forge/linux-64::pennylane-0.29.0-py310hff52083_1 
  pennylane-core     conda-forge/linux-64::pennylane-core-0.29.0-py310hff52083_1 
  pennylane-lightni~ conda-forge/linux-64::pennylane-lightning-core-0.29.0-py310h4f277b2_1 
...
```
but it's kinda ugly...

**Comment by vincentmr** [comment]
> We can certainly add a version constraint like I mentioned https://github.com/conda-forge/pennylane-lightning-gpu-feedstock/issues/1#issuecomment-1496089447:

Yes, at least what definitely should work, works. But I agree the default should be the latest version.

**Comment by leofang** [comment]
> As promised, let's get M1/M2 supported in a separate PR.

So I submitted https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4302, and after merging the bot sent #20 for osx-arm64 and #19 for aarch64/ppc64le.

**Comment by BastianZim** [comment]
@leofang Would you like to be a member as well since you're the resident CUDA expert. :)

**Comment by vincentmr** [comment]
Hi @BastianZim , would you have some time to look at my [Kokkos-kernels PR](https://github.com/conda-forge/staged-recipes/pull/22345)? After it's merged, we could move on to adding the Kokkos and Kokkos-kernels dependencies back into the build as we discussed.

**Comment by leofang** [comment]
> @leofang Would you like to be a member as well since you're the resident CUDA expert. :)

This feedstock does not need any CUDA expert 😄 But it seems I better add myself to fix #20...

**Pull request merged**